### PR TITLE
envision: init at 0-unstable-2024-06-23

### LINKED ISSUE #292628: Package request: envision
**Project description**

UI for building, configuring and running Monado, the open source OpenXR runtime.

**Metadata**

* homepage URL: https://gitlab.com/gabmus/envision
* source URL: https://gitlab.com/gabmus/envision
* license: agpl3
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Pandapip1**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/292628

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/en/envision/package.nix
Comment (superherointj) [inline pkgs/by-name/en/envision/package.nix:43]: If there isn't a version in upstream, it should be `0-unstable-2024-06-16`.

**Comment by Pandapip1** [comment]
Blocked by https://github.com/NixOS/nixpkgs/pull/316975

**Comment by Pandapip1** [comment]
Gave up trying to get it to work without fhsenv.

File: pkgs/by-name/en/envision/envision-unwrapped.nix
Comment (Pandapip1) [inline pkgs/by-name/en/envision/envision-unwrapped.nix:83]: CC @Scrumplex. Would you like to be listed as a maintainer?

Comment (Scrumplex) [reply pkgs/by-name/en/envision/envision-unwrapped.nix:83]: yup ^^

**Comment by Pandapip1** [comment]
Updated to latest unstable.

File: pkgs/by-name/en/envision-unwrapped/package.nix
Comment (superherointj) [inline pkgs/by-name/en/envision-unwrapped/package.nix:97]:
```suggestion
  meta = {
    description = "UI for building, configuring and running Monado, the open source OpenXR runtime";
    homepage = "https://gitlab.com/gabmus/envision";
    license = lib.licenses.agpl3Only;
    mainProgram = "envision";
    maintainers = with lib.maintainers; [ pandapip1 Scrumplex ];
    platforms = lib.platforms.linux;
  };
```
Avoiding [unnecessary](https://nix.dev/guides/best-practices.html#with-scopes) nested with.

**Comment by Pandapip1** [comment]
> Nix files are to be formatted with [nixfmt-rfc-style](https://search.nixos.org/packages?query=nixfmt).

Why is there not CI for this?

**Comment by superherointj** [comment]
> > Nix files are to be formatted with [nixfmt-rfc-style](https://search.nixos.org/packages?query=nixfmt).
> 
> Why is there not CI for this?

Files not listed in:

https://github.com/NixOS/nixpkgs/blob/master/.github/workflows/check-nix-format.yml#L39

Once formatted, it can be enable CI formatting there.

File: pkgs/by-name/en/envision/package.nix
Comment (superherointj) [inline pkgs/by-name/en/envision/package.nix:3]: nixfmt-rfc-style ^

File: nixos/modules/programs/envision.nix
Comment (superherointj) [inline nixos/modules/programs/envision.nix:54]:
```suggestion
  meta.maintainers = with lib.maintainers; [ pandapip1 ];
```
It's possible to reference the `package.meta.maintainers` here. And save having to edit many files.

File: nixos/modules/programs/envision.nix
Comment (superherointj) [inline nixos/modules/programs/envision.nix:40]: Are you sure this should be `root` here? Isn't this a GUI application? Why root access?

Comment (superherointj) [reply nixos/modules/programs/envision.nix:40]: @wegank Thoughts on this?

Comment (Pandapip1) [reply nixos/modules/programs/envision.nix:40]: `setuid` is false. The file has to have an owner, so `root` seems to be a sensible default that's hard to edit maliciously.

Comment (wegank) [reply nixos/modules/programs/envision.nix:40]: I guess it's up to the author of the PR to explain, I have no idea what permission envision needs...

**Comment by Pandapip1** [comment]
Just started testing out the module. I forgot that bwrap doesn't like capabilities. https://github.com/NixOS/nixpkgs/pull/309906 will need to be merged first.

**Comment by Pandapip1** [comment]
Never mind. I'll just not have the security wrapper then.

**Comment by Pandapip1** [comment]
Did another test. WiVRn is definitely choppy, but it works.

Stuff that doesn't work:

- Root permission prompts for adding capabilities
- WiVRn installation through Envision (`adb install ~/.cache/envision/wivrn.apk` works fine)

**Comment by superherointj** [comment]
The commit message for NixOS module is out of standard:
https://github.com/NixOS/nixpkgs/blob/master/nixos/README.md#commit-conventions

> nixos/envision: init

The one for the docs, I suggest:
> nixos/doc/rl-2411: add programs.envision

**Comment by Pandapip1** [comment]
Ah, it was in the README, not the CONTRIBUTING document. I was looking for that!

**Review by superherointj — APPROVED** [review]

**Pull request merged**
